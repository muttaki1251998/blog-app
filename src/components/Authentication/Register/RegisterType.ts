export interface RegisterType {
    username: string;
    email: string;
    password: string;
}

export interface ErrorType {
    usernameError: string | null;
    emailError: string | null;
    passwordError: string | null;
}