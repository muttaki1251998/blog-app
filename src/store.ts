import {configureStore} from '@reduxjs/toolkit';
import { useDispatch } from 'react-redux';
import authReducer from './components/Authentication/authSlice';

export const store = configureStore({
    reducer: authReducer
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch
export const useAppDispatch = useDispatch.withTypes<AppDispatch>() //
