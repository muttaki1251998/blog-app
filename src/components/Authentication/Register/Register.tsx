import React, { useState, ChangeEvent, useCallback } from "react";
import axios from "axios";
import { useDispatch } from "react-redux";
import { register } from "../../../services/authServices/authService";
import { RegisterType, ErrorType } from "./RegisterType";
import { AppDispatch } from "../../../store";
import { debounce } from "lodash";

const Register: React.FC = () => {
  const [username, setUsername] = useState<string>("");
  const [email, setEmail] = useState<string>("");
  const [password, setPassword] = useState<string>("");
  const initialErrorStates: ErrorType = {
    usernameError: null,
    emailError: null,
    passwordError: null,
  };
  const [error, setError] = useState<ErrorType>(initialErrorStates);

  const dispatch = useDispatch<AppDispatch>();

  const validationSchema = (data: RegisterType) => {
    const { username, email, password } = data;
    const validEmail =
      /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;

    // Username validation
    if (username.length === 0) {
      setError((prev) => ({
        ...prev,
        usernameError: "This field cannot be empty",
      }));
    } else if (username.length < 4) {
      setError((prev) => ({
        ...prev,
        usernameError: "Username must have atleast 4 letters",
      }));
    } else {
      setError((prev) => ({ ...prev, usernameError: null }));
    }

    // Email validation
    if (email.length === 0) {
      setError((prev) => ({
        ...prev,
        emailError: "This field cannot be empty",
      }));
    } else if (email.length < 4) {
      setError((prev) => ({ ...prev, emailError: "Email too short" }));
    } else if (!email.match(validEmail)) {
      setError((prev) => ({ ...prev, emailError: "Invalid Email" }));
    } else {
      setError((prev) => ({ ...prev, emailError: null }));
    }

    // Password validation
    if (password.length === 0) {
      setError((prev) => ({
        ...prev,
        passwordError: "This field cannot be empty",
      }));
    } else if (password.length < 4) {
      setError((prev) => ({ ...prev, passwordError: "Password too short" }));
    } else {
      setError((prev) => ({ ...prev, passwordError: null }));
    }
  };
  const debounceValidation = useCallback(debounce(validationSchema, 300), []);
  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    const { id, value } = e.target;
    if (id === "username") setUsername(value);
    else if (id === "email") setEmail(value);
    else if (id === "password") setPassword(value);

    debounceValidation({ username, email, password });
  };

  const handleSubmit = async (e: any) => {
    e.preventDefault();
    const user: RegisterType = {
      username,
      email,
      password,
    };
    validationSchema(user);
    if (Object.values(error).every((value) => value !== null)) {
      dispatch(register(user));
    }
  };


  return (
    <form
      onSubmit={(e) => handleSubmit(e)}
      className="flex items-center justify-center bg-gray-200"
      style={{ height: "calc(100vh - 64px)", overflowY: "auto" }}
    >
      <div className="w-full max-w-md">
        <label className="text-lg text-gray-800 text-center block mb-4">
          Join us!
        </label>
        <div className="mb-4 w-full">
          <input
            type="text"
            id="username"
            data-testid="username"
            value={username}
            name="username"
            onChange={(e) => handleChange(e)}
            placeholder="Username"
            className={`shadow border appearance-none ${
              error.usernameError ? "border-red-400" : "border-blue-400"
            } rounded text-gray-700 focus:outline-blue-400 w-full py-2 px-4`}
          />
          <div>
            {error.usernameError && (
              <p className="text-red-700 text-sm">{error.usernameError}</p>
            )}
          </div>
        </div>
        <div className="mb-4">
          <input
            type="text"
            id="email"
            value={email}
            name="email"
            onChange={(e) => handleChange(e)}
            placeholder="Email"
            className={`shadow border appearance-none ${
              error.emailError ? "border-red-400" : "border-blue-400"
            } rounded text-gray-700 focus:outline-blue-400 w-full py-2 px-4`}
          />
          <div>
            {error.emailError && (
              <p className="text-red-700 text-sm">{error.emailError}</p>
            )}
          </div>
        </div>
        <div className="mb-4">
          <input
            type="password"
            id="password"
            value={password}
            name="password"
            onChange={(e) => handleChange(e)}
            placeholder="Password"
            className={`shadow border appearance-none ${
              error.passwordError ? "border-red-400" : "border-blue-400"
            } rounded text-gray-700 focus:outline-blue-400 w-full py-2 px-4`}
          />
          <div>
            {error.passwordError && (
              <p className="text-red-700 text-sm">{error.passwordError}</p>
            )}
          </div>
        </div>
        <button
          type="submit"
          className="bg-black rounded shadow py-2 px-4 text-white justify-center hover:bg-gray-700 transition duration 300 ease-in-out"
        >
          Submit
        </button>
      </div>
    </form>
  );
};

export default Register;
