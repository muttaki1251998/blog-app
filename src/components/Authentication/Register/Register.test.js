import React from "react";
import "@testing-library/jest-dom";
import { fireEvent, screen, render } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../../../store";
import Register from "./Register";

describe("snapshot", () => {
  it("should pass snapshot tests", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Register />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

// describe("component tests", () => {
//   it("should show error message when username is short", async () => {
//     render(
//       <Provider store={store}>
//         <Register />
//       </Provider>
//     );
//     const usernameInput = screen.getByTestId("username");
//     fireEvent.change(usernameInput, { target: { value: "a" } });
//     fireEvent.blur(usernameInput);

//     expect(
//       await screen.findByText("This field cannot be empty", {}, {timeout: 400})
//     ).toBeInTheDocument();
//   });
// });
