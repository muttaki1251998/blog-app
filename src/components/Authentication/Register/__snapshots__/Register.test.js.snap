// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot should pass snapshot tests 1`] = `
<DocumentFragment>
  <form
    class="flex items-center justify-center bg-gray-200"
    style="height: calc(100vh - 64px); overflow-y: auto;"
  >
    <div
      class="w-full max-w-md"
    >
      <label
        class="text-lg text-gray-800 text-center block mb-4"
      >
        Join us!
      </label>
      <div
        class="mb-4 w-full"
      >
        <input
          class="shadow border appearance-none border-blue-400 rounded text-gray-700 focus:outline-blue-400 w-full py-2 px-4"
          data-testid="username"
          id="username"
          name="username"
          placeholder="Username"
          type="text"
          value=""
        />
        <div />
      </div>
      <div
        class="mb-4"
      >
        <input
          class="shadow border appearance-none border-blue-400 rounded text-gray-700 focus:outline-blue-400 w-full py-2 px-4"
          id="email"
          name="email"
          placeholder="Email"
          type="text"
          value=""
        />
        <div />
      </div>
      <div
        class="mb-4"
      >
        <input
          class="shadow border appearance-none border-blue-400 rounded text-gray-700 focus:outline-blue-400 w-full py-2 px-4"
          id="password"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
        <div />
      </div>
      <button
        class="bg-black rounded shadow py-2 px-4 text-white justify-center hover:bg-gray-700 transition duration 300 ease-in-out"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</DocumentFragment>
`;
